File: transports/csharp/src/Csilgen.Transport/Datagrams.cs
// CSIL-Datagrams transport -- unreliable, unordered, message-oriented -- see
// csil-datagrams-transport.md. CBOR-array (default) and compact fixed-header profiles. A datagram
// channel is single-service: the service is bound at channel setup, so datagrams carry no service
// ordinal.

namespace Csilgen.Transport;

public static class Datagrams
{
    // MaxDatagramDefault is the conservative max datagram size (envelope + payload) safe across
    // UDP/WebRTC/QUIC.
    public const int MaxDatagramDefault = 1200;
}

// Datagram is a datagram in the CBOR-array (default) profile: [v, op_ord, seq, payload]. Seq is the
// per-channel sequence; 0 means "unsequenced". Payload is the opaque CBOR(message type) bytes.
public sealed record Datagram(ulong OpOrd, ulong Seq, byte[] Payload)
{
    public static Datagram New(ulong opOrd, ulong seq, byte[] payload) => new(opOrd, seq, payload);

    public byte[] Encode()
    {
        var arr = new CborArray(new List<CborValue>
        {
            new CborUint(Conventions.Version),
            new CborUint(OpOrd),
            new CborUint(Seq),
            Conventions.Tag24(Payload),
        });
        return Cbor.Encode(arr);
    }

    public static Datagram Decode(ReadOnlySpan<byte> bytes)
    {
        CborValue v = Cbor.DecodeEnvelope(bytes);
        if (v is not CborArray arr)
        {
            throw new MalformedException("datagram is not an array");
        }

        if (arr.Items.Count != 4)
        {
            throw new MalformedException(
                $"datagram array has {arr.Items.Count} elements, expected 4");
        }

        if (!Conventions.TryAsU64(arr.Items[0], out ulong ver))
        {
            throw new MalformedException("datagram field not an integer");
        }

        Conventions.CheckVersion(ver);
        if (!Conventions.TryAsU64(arr.Items[1], out ulong opOrd)
            || !Conventions.TryAsU64(arr.Items[2], out ulong seq))
        {
            throw new MalformedException("datagram field not an integer");
        }

        return new Datagram(opOrd, seq, Conventions.Untag24(arr.Items[3]));
    }
}

// CompactDatagram is a datagram in the compact fixed-header profile. Header layout:
// [ver|flags][op_ord:u8][seq:u16 BE]([epoch:u8]) then the opaque body. Epoch is present when the
// sender tracks restarts (sets the flags epoch bit). Body is opaque (tag-24 CBOR or a raw media
// frame, by channel agreement).
public sealed record CompactDatagram(byte OpOrd, ushort Seq, byte[] Body)
{
    private const byte CompactVer = 1;
    private const byte FlagEpoch = 0b0001;

    public byte? Epoch { get; init; }

    public static CompactDatagram New(byte opOrd, ushort seq, byte[] body) =>
        new(opOrd, seq, body);

    // WithEpoch sets the epoch byte (and the flags bit that signals its presence).
    public CompactDatagram WithEpoch(byte epoch) => this with { Epoch = epoch };

    public byte[] Encode()
    {
        byte flags = Epoch is not null ? FlagEpoch : (byte)0;
        var outBytes = new List<byte>(5 + Body.Length)
        {
            (byte)((CompactVer << 4) | (flags & 0x0f)),
            OpOrd,
            (byte)(Seq >> 8),
            (byte)Seq,
        };
        if (Epoch is byte epoch)
        {
            outBytes.Add(epoch);
        }

        outBytes.AddRange(Body);
        return outBytes.ToArray();
    }

    public static CompactDatagram Decode(ReadOnlySpan<byte> b)
    {
        if (b.Length < 4)
        {
            throw new MalformedException("compact datagram shorter than the 4-byte header");
        }

        byte ver = (byte)(b[0] >> 4);
        if (ver != CompactVer)
        {
            throw new UnsupportedVersionException(ver);
        }

        byte flags = (byte)(b[0] & 0x0f);
        byte opOrd = b[1];
        ushort seq = (ushort)((b[2] << 8) | b[3]);
        byte? epoch = null;
        int bodyStart = 4;
        if ((flags & FlagEpoch) != 0)
        {
            if (b.Length < 5)
            {
                throw new MalformedException(
                    "compact datagram flags claim an epoch byte that is absent");
            }

            epoch = b[4];
            bodyStart = 5;
        }

        return new CompactDatagram(opOrd, seq, b[bodyStart..].ToArray()) { Epoch = epoch };
    }
}

// SeqEventKind classifies an incoming sequence number relative to what was last seen, for
// loss/reorder/restart detection. The transport detects; the app decides.
public enum SeqEventKind
{
    // First is the first datagram seen on the channel.
    First,

    // Advanced is strictly newer than the last (possibly skipping some -- a gap/loss).
    Advanced,

    // LateOrDuplicate is not newer (a late or duplicate datagram).
    LateOrDuplicate,

    // Restart means the sender restarted (epoch changed); seq numbering reset.
    Restart,
}

// SeqEvent is a sequence classification plus the gap count for Advanced.
public readonly record struct SeqEvent(SeqEventKind Kind, ulong Gap = 0);

// SeqTracker tracks the last sequence/epoch per channel to classify arrivals. Unsequenced datagrams
// (seq 0) are reported as Advanced with gap 0.
public sealed class SeqTracker
{
    private ulong? _lastSeq;
    private byte? _lastEpoch;

    // Observe classifies an arriving (seq, epoch) and updates the tracker state.
    public SeqEvent Observe(ulong seq, byte? epoch)
    {
        // A restart fires only when a prior epoch existed and changed; no-epoch -> first epoch is
        // not a restart.
        if (!EpochEqual(epoch, _lastEpoch) && _lastEpoch is not null)
        {
            _lastEpoch = epoch;
            _lastSeq = seq;
            return new SeqEvent(SeqEventKind.Restart);
        }

        _lastEpoch = epoch;

        // seq 0 marks an unsequenced datagram: it carries no ordering information, so it is never
        // late or duplicate. Report a zero-gap advance and leave the running sequence untouched so a
        // mix of sequenced and unsequenced still tracks the sequenced ones.
        if (seq == 0)
        {
            return new SeqEvent(SeqEventKind.Advanced, 0);
        }

        if (_lastSeq is not ulong last)
        {
            _lastSeq = seq;
            return new SeqEvent(SeqEventKind.First);
        }

        if (seq > last)
        {
            ulong gap = seq - last - 1;
            _lastSeq = seq;
            return new SeqEvent(SeqEventKind.Advanced, gap);
        }

        return new SeqEvent(SeqEventKind.LateOrDuplicate);
    }

    private static bool EpochEqual(byte? a, byte? b)
    {
        if (a is null || b is null)
        {
            return a is null && b is null;
        }

        return a.Value == b.Value;
    }
}

File: transports/swift/Sources/CsilgenTransport/Datagrams.swift
// CSIL-Datagrams transport — unreliable, unordered, message-oriented — see
// csil-datagrams-transport.md. CBOR-array (default) and compact fixed-header profiles.
// A datagram channel is single-service: the service is bound at channel setup, so
// datagrams carry no service ordinal.

/// A datagram in the CBOR-array (default) profile: `[v, op_ord, seq, payload]`. Every
/// field is a full CBOR uint here (unlike the compact profile's fixed widths), so the
/// op_ord/seq carry no width ceiling and the payload is the usual tag-24 byte string.
public struct Datagram: Equatable, Sendable {
    public var opOrd: UInt64
    /// The per-channel sequence; 0 means "unsequenced".
    public var seq: UInt64
    /// The opaque CBOR(message type) bytes (wrapped in tag 24 on the wire).
    public var payload: [UInt8]

    public init(opOrd: UInt64, seq: UInt64, payload: [UInt8]) {
        self.opOrd = opOrd
        self.seq = seq
        self.payload = payload
    }

    public func encode() -> [UInt8] {
        encodeValue(.array([.uint(csilVersion), .uint(opOrd), .uint(seq), tag24(payload)]))
    }

    public static func decode(_ b: [UInt8]) throws -> Datagram {
        let v = try decodeEnvelope(b)
        guard case .array(let arr) = v else { throw malformed("datagram is not an array") }
        guard arr.count == 4 else {
            throw malformed("datagram array has \(arr.count) elements, expected 4")
        }
        guard let ver = asU64(arr[0]), let opOrd = asU64(arr[1]), let seq = asU64(arr[2]) else {
            throw malformed("datagram field not an integer")
        }
        try checkVersion(ver)
        return Datagram(opOrd: opOrd, seq: seq, payload: try untag24(arr[3]))
    }
}

/// A datagram in the compact fixed-header profile. The header is a hand-packed binary
/// layout (NOT CBOR) so the hot media path never parses CBOR for the header:
/// `[ver|flags][op_ord:u8][seq:u16 BE]([epoch:u8])` then the opaque body. The fixed
/// widths come from csil-datagrams-transport.md §2.2, not from the conformance JSON
/// (which leaves its numbers untyped).
public struct CompactDatagram: Equatable, Sendable {
    public var opOrd: UInt8
    public var seq: UInt16
    /// Present when the sender tracks restarts (sets the flags epoch bit).
    public var epoch: UInt8?
    /// The opaque body bytes (tag-24 CBOR or a raw media frame, by channel agreement).
    public var body: [UInt8]

    public init(opOrd: UInt8, seq: UInt16, body: [UInt8], epoch: UInt8? = nil) {
        self.opOrd = opOrd
        self.seq = seq
        self.body = body
        self.epoch = epoch
    }

    /// Set the epoch byte (and the flags bit that signals its presence).
    public func withEpoch(_ epoch: UInt8) -> CompactDatagram {
        var copy = self
        copy.epoch = epoch
        return copy
    }

    private static let compactVer: UInt8 = 1
    private static let flagEpoch: UInt8 = 0b0001

    public func encode() -> [UInt8] {
        let flags: UInt8 = epoch != nil ? Self.flagEpoch : 0
        var out: [UInt8] = []
        out.reserveCapacity(5 + body.count)
        out.append((Self.compactVer << 4) | (flags & 0x0f))
        out.append(opOrd)
        out.append(UInt8(seq >> 8))
        out.append(UInt8(seq & 0xff))
        if let epoch { out.append(epoch) }
        out.append(contentsOf: body)
        return out
    }

    public static func decode(_ b: [UInt8]) throws -> CompactDatagram {
        guard b.count >= 4 else {
            throw malformed("compact datagram shorter than the 4-byte header")
        }
        let ver = b[0] >> 4
        guard ver == compactVer else { throw TransportError.unsupportedVersion(UInt64(ver)) }
        let flags = b[0] & 0x0f
        let opOrd = b[1]
        let seq = UInt16(b[2]) << 8 | UInt16(b[3])
        var epoch: UInt8? = nil
        var bodyStart = 4
        if flags & flagEpoch != 0 {
            guard b.count >= 5 else {
                throw malformed("compact datagram flags claim an epoch byte that is absent")
            }
            epoch = b[4]
            bodyStart = 5
        }
        return CompactDatagram(
            opOrd: opOrd, seq: seq, body: Array(b[bodyStart...]), epoch: epoch)
    }
}

/// Classifies an incoming sequence number relative to what was last seen, for
/// loss/reorder/restart detection. The transport detects; the app decides policy.
public enum SeqEventKind: Equatable, Sendable {
    /// The first datagram seen on the channel.
    case first
    /// Strictly newer than the last (possibly skipping some — a gap/loss), with the
    /// count of skipped sequence numbers.
    case advanced(gap: UInt64)
    /// Not newer than the last seen (a late or duplicate datagram).
    case lateOrDuplicate
    /// The sender restarted (epoch changed); seq numbering reset.
    case restart
}

/// Tracks the last sequence/epoch per channel to classify arrivals. Unsequenced
/// datagrams (seq 0) are reported as `.advanced(gap: 0)`.
public final class SeqTracker {
    private var lastSeq: UInt64?
    private var lastEpoch: UInt8?

    public init() {}

    /// Classify an arriving (seq, epoch) and update the tracker state.
    public func observe(seq: UInt64, epoch: UInt8?) -> SeqEventKind {
        // A restart fires only when a *prior* epoch existed and changed; going from
        // no-epoch to a first epoch is not a restart.
        if epoch != lastEpoch, lastEpoch != nil {
            lastEpoch = epoch
            lastSeq = seq
            return .restart
        }
        lastEpoch = epoch
        // seq 0 marks an unsequenced datagram: it carries no ordering information, so it
        // is never late or duplicate. Report a zero-gap advance and leave the running
        // sequence untouched so a mix of sequenced and unsequenced still tracks the
        // sequenced ones.
        if seq == 0 {
            return .advanced(gap: 0)
        }
        guard let last = lastSeq else {
            lastSeq = seq
            return .first
        }
        if seq > last {
            lastSeq = seq
            return .advanced(gap: seq - last - 1)
        }
        return .lateOrDuplicate
    }
}
